doc/developer/design/20210311_prometheus_source.md

```diff
@@ -0,0 +1,144 @@
+
+```sql
+CREATE SOURCE metrics_table_name
+PROMETHEUS ENDPOINT 'http://localhost:9090/metrics'
+WITH (
+    metadata_table = "metrics_table_meta",
+    poll_interval = '1 minute'::interval,
+    retain_data = '5 minutes'::interval
+);
+```
+
+This creates two relations that have schemas similar to the following:
+
+```sql
+CREATE TABLE metrics_table_name (
+    metric TEXT,
+    ts TIMESTAMP,
+    labels JSONB,
+    value float8
+);
+
+CREATE TABLE metrics_table_meta (
+    metric TEXT,
+    metric_type TEXT,
+    help TEXT
+);
+```
+
+Once created, the source polls the endpoint once a minute (via the configured the `poll_interval` duration), and retains metric readings for 5 minutes.
+
+
+# Alternatives
+
+<!-- // Similar to the Description section. List of alternative approaches considered, pros/cons or why they were not chosen -->
+
+There are two major things that we could do differently to fulfill the brief, and a series of things that we could do to polish this design. I'll summarize the former under "Alternative approaches" and the latter under "Alternative details".
+
+## Alternative approaches
+
+### Create an internal metrics reporting channel instead (feeding metrics directly)
+
+We could add a mechanism for providing log events to the dataflow crate (or a crate that df and others depend on), where if a piece of code now increments a prometheus counter, it would instead send an event that gets captured by a listener on a channel, and increments the prometheus cuonter, as well as queue the event for updating of internal metrics views.
+
+The major benefit is that we encapsulate our metrics tooling usage more (making it easier to plug in more ways of recording metric values).
+
+The downsides, however, are somewhat big:
+
+* Things need to be structured such that you can get at the logger where you need to count the metric. Some parts where we currently do this make this somewhat difficult (e.g. the pgwire code, where the currently best way to count bytes sent is in encoding).
+
+* Adding metrics gets more difficult. You not only have to figure out how go get your number counted the right way, you also have to figure out how to aggregate it properly in the dataflow world. We worry that this might mean that useful operational metrics don't get added as soon/as quickly as they'd be necessary, making it harder to run the system.
+
+### Do not scrape arbitrary prometheus endpoints, instead scrape our internal metrics registry
+
+This would get around the need to introduce an externally-definable source, and automatically feed a dataflow that we can keep up to date automatically.
+
+There's one major reason why I think this would be harder to implement right now: Being able to pull in all our metrics depends on [#5825], which would be a pretty heavy re-work of our internal structures (have to thread a registry through to various bits that don't have a place to hold that registry yet, like the pgwire implementation).
```

Comment:
This isn't correct: we can 100% get all metrics data into our system tables without that issue, if anything doing it before we finish #5825 will be microscopically easier because we won't need to pass the registry into 1 more place.

Reply:
@benesch has voiced the same point, and thinking about this more - I agree! Now that I think about this, the causality in the paragraph above is the wrong way around: scraping of the metrics registry would make it microscopically harder to fix #5825, much like adding any single metric does (-:

So, given these points (by you & Nikhil), I feel that we should not make an externally-scraping source yet but instead go with this design: Feeding a dataflow from the global (or, later, the "materialized") registry and skipping the "poll a random prometheus endpoint" bits.